src/dataunion/DataUnion.ts

```diff
@@ -564,30 +590,109 @@ export class DataUnion {
+
+        log('Executing DataUnionSidechain withdraw function')
         const tx = await getWithdrawTxFunc()
         const tr = await tx.wait()
-        if (!freeWithdraw && sendToMainnet) {
-            await this.getContracts().transportSignaturesForTransaction(tr, options)
+
+        // keep tokens in the sidechain => just return the sidechain tx receipt
+        if (!sendToMainnet) { return tr }
+
+        log(`Got receipt, filtering UserRequestForSignature from ${tr.events!.length} events...`)
+        const ambHashes = getMessageHashes(tr)
+
+        if (ambHashes.length < 1) {
+            throw new Error("No UserRequestForSignature events emitted from withdraw transaction, can't transport withdraw to mainnet")
+        }
+
+        if (ambHashes.length > 1) {
+            throw new Error('Expected only one UserRequestForSignature event')
+        }
+
+        const messageHash = ambHashes[0]
+
+        if (!payForTransport) {
+            // expect someone else to do the transport for us (corresponds to dataUnion.freeWithdraw=true)
+            if (waitUntilTransportIsComplete) {
+                log(`Waiting for balance=${balanceBefore.toString()} change (poll every ${pollingIntervalMs}ms, timeout after ${retryTimeoutMs}ms)`)
+                await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs).catch((e) => {
+                    const msg = `Timeout: Bridge did not transport withdraw message as expected. Fix: DataUnion.transportMessage("${messageHash}")`
+                    throw e.message.startsWith('Timeout') ? new Error(msg) : e
+                })
+                return null
+            }
+
+            // instead of waiting, hand out the messageHash so that someone else might do the transport using it
+            return messageHash
+        }
+
+        const ambTr = await this.transportMessage(messageHash, pollingIntervalMs, retryTimeoutMs)
+        if (waitUntilTransportIsComplete) {
+            log(`Waiting for balance ${balanceBefore.toString()} to change`)
+            await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs)
+        }
+        return ambTr
+    }
+
+    // TODO: this doesn't belong here. Transporting a message is NOT dataunion-specific and needs nothing from DataUnion.ts.
+    //       It shouldn't be required to create a DataUnion object to call this.
+    //       This belongs to the StreamrClient, and if the code is too DU-specific then please shove it back to Contracts.ts.
+    //       Division to transportMessage and Contracts.transportSignaturesForMessage is spurious, they should be one long function probably.
+    /**
+     * @returns null if message was already transported, ELSE the mainnet AMB signature execution transaction receipt
+     */
+    async transportMessage(messageHash: AmbMessageHash, pollingIntervalMs: number = 1000, retryTimeoutMs: number = 300000) {
```

Comment:
`retryTimeoutMs: number = 300000`

5 minutes seems like a long time

Reply:
yea but we don't know how the bridge oracles operate. 5 minutes might in fact be what it may take in production. It's not great for web sites obviously. Maybe there needs to be a more multi-step process where you can add a listener for when it's ready or smth (e.g. get back the message hash anyway, store it to localstorage, then start listening to when it's transported; if page gets refreshed or something, get that "pending hash" from localstorage and resume listening)

None of that is something I've planned as part of this ticket. I'll make a new ticket, but the feature is going to be for the front-end people, so I can't really say how urgent it would be. It would change the public interface though...

Reply:
https://linear.app/streamr/issue/ETH-102/split-withdraw-with-signature-transport-more-frontend-ergonomically